Refine side UI

jasmussen opened the pull request:
This PR optimizes and refines the side UI. In doing so, it fixes #5400 and #4223 by making the arrows visually bigger. It also simplifies the hover styles and improves mobile in the process.

These are now essentially "icon buttons", in that they have the same hover, click and border radius style. Although the buttons are only slightly bigger, the visually bigger hit area makes it _feel_ more clickable as well.

![refine side ui](https://user-images.githubusercontent.com/1204802/38673180-d385e57a-3e50-11e8-958e-f25bca69c21c.gif)

As you can see, I also snuck in a trash button. Let me know your thoughts on that — it's easy to remove and restore later on, if we need to do that to ship the other enhancements, but it's very nice to have it surfaced there, and brings with it a nice consistency with mobile.

Commit:
Polish side UI, make arrows bigger.

This is initial work to improve the side UI. It fixes #5400 and #4223 by making the arrows visually bigger. It also simplifies the hover styles and improves mobile in the process.

Commit:
Refine mobile block UI, and responsive hover labels.

Commit:
Add trash button to right side UI.

More to come.

Commit:
Enhance situation in wide blocks.

chrisvanpatten commented:
This looks like a huge improvement although the rounded border radius bumping right up against the block border feels a bit off.

![refine_side_ui_by_jasmussen_ _pull_request__6141_ _wordpress_gutenberg](https://user-images.githubusercontent.com/1231306/38676263-74116538-3e28-11e8-9dc8-0ccedeeb903c.png)

I couldn't figure out a way to do it perfectly in code (you'd probably need to apply the border styles to the SVG, or a new nested child element?) but I mocked up how it might look:

![_untitled____ _paleo_meal_plans_ _wordpress](https://user-images.githubusercontent.com/1231306/38678706-c7b05328-3e2f-11e8-8d89-0d3133afac4c.png)

That extra bit of breathing space feels good, and if it's treated as clickable area on the button it should keep it easy to hit.

mtias commented on editor/components/block-settings-menu/index.js:
```diff
@@ -84,8 +84,9 @@ export class BlockSettingsMenu extends Component {
-						/>
-					);
+						/>,
+						<BlockRemoveButton key="remove" uids={ uids } role="menuitem" />
```
I think the role should be removed here.

SuperGeniusZeb commented:
Personally, I am glad to see the trash button return. It is pretty handy, and I do not think it adds much weight to the UI, especially with the changes in #6101. These new styles look pretty nice and do feel more clickable, which is good. I like @chrisvanpatten's suggestion of having a little gap between them and the visible block outline.

Commit:
Remove role from Remove button, and remove Trash from menu.

gziolo committed:
Editor: Fix Eslint issues in BlockSettingsMenu

gziolo commented:
Great improvements, I have a similar concern regarding this very subtle thing with overlapping borders:

![screen shot 2018-04-13 at 11 06 35](https://user-images.githubusercontent.com/699132/38726598-e28c72cc-3f0a-11e8-9a4e-397e6cf59ef9.png)

Commit:
Refine collapsing margins hack, fix flickering.

Commit:
Add some clearance between side UI and block.

jasmussen commented:
Fixed the flickering issue that made things look slightly cropped in weird circumstances, and added a little clearance as suggested by @chrisvanpatten:

![side ui](https://user-images.githubusercontent.com/1204802/38730674-c8b4f7d6-3f17-11e8-8e41-c170b1ee75f1.gif)

Thanks for the reviews folks, good to go?

gziolo commented:
I noticed regression on mobile caused by the fact that icons are swapped on the right side.

### Before

![screen shot 2018-04-13 at 14 41 47](https://user-images.githubusercontent.com/699132/38735370-eec3ae48-3f28-11e8-8370-491a1397e846.png)
![screen shot 2018-04-13 at 14 42 00](https://user-images.githubusercontent.com/699132/38735371-eeea7bd6-3f28-11e8-950c-e24416832812.png)

### After

![screen shot 2018-04-13 at 14 38 06](https://user-images.githubusercontent.com/699132/38735391-f59bef5a-3f28-11e8-983c-d479069d5c2e.png)
![screen shot 2018-04-13 at 14 38 28](https://user-images.githubusercontent.com/699132/38735393-f5ba5382-3f28-11e8-90e5-2aa3606e03d8.png)


The issues with `z-index` is unrelated to this PR, but nice to fix anyways :)

Commit:
Fix z-index issue.

jasmussen commented:
I pushed a fix for the z-index, however I would appreciate some help with the dropdown menu. Turns out there are issues on all the breakpoints due to the addition of the BlockRemove button:

<img width="150" alt="screen shot 2018-04-16 at 08 45 52" src="https://user-images.githubusercontent.com/1204802/38793754-c94fcb02-4152-11e8-8d3a-dc9f5b8c2233.png">

<img width="539" alt="screen shot 2018-04-16 at 08 46 26" src="https://user-images.githubusercontent.com/1204802/38793761-cd2d9380-4152-11e8-9cf2-0dacbbe3e5e5.png">

<img width="395" alt="screen shot 2018-04-16 at 08 46 35" src="https://user-images.githubusercontent.com/1204802/38793765-d0248d28-4152-11e8-94e8-bcba80050176.png">

@gziolo if you have bandwidth to look at it, I'd really appreciate it, thank you.

youknowriad committed:
Fix block settings dropdown menu

jasmussen commented:
Thanks so much Riad! Okay _now_, thanks to Riad, this is ready for final review.

jorgefilipecosta commented:
It looks like there we a small white line at the side of the mover (noticed when the background is not white):
![image](https://user-images.githubusercontent.com/11271197/38870627-4dc26612-4246-11e8-9c83-51a32403be81.png).

jorgefilipecosta commented on editor/components/block-mover/style.scss:
```diff
@@ -17,14 +19,21 @@
+		padding: ( $block-side-ui-width - 18px ) / 2;	// this makes the SVG fill the whole available area, without scaling the artwork
+	}
+
+	// Apply a background in nested contexts, only on desktop
```
In nested contexts on mobile the contrast may be low:
![image](https://user-images.githubusercontent.com/11271197/38870696-87c8c4be-4246-11e8-869b-3ad7736fdb50.png). But it seems like the result is better than what we had and finding a solution for this cases is really hard has the background color of the nested context is impossible to predict.

jorgefilipecosta commented:
Changes here improve the look and feel in nested contexts and the PR  https://github.com/WordPress/gutenberg/pull/5452 will benefit a lot from this changes 👍 Thank you @jasmussen!

jasmussen commented:
Thanks for the review, @jorgefilipecosta!

Your feedback on other colored backgrounds is duly noted, and applies not only to nested contexts, but presumably also to when themes load their own stylesheets into the editor. It is on my todo list to tackle this next — perhaps using mix-blend-mode as @lsl suggested, or just replaing the grays with blacks and whites with varying opacities to better fit in, while maintaining contrast.

Do you think, perhaps, that we should merge this PR in as is, and then tackle this separately, though? It seems like it would be nice to get these changes into 2.7 if we can.

jasmussen commented:
Got a thumbs up in chat. Going to merge this as is, and then tackle UI contrast issues separately.
